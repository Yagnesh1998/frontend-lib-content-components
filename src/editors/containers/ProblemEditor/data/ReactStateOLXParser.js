import _ from 'lodash-es';
import { XMLParser, XMLBuilder } from 'fast-xml-parser';
import { ProblemTypeKeys } from '../../../data/constants/problem';

class ReactStateOLXParser {
  constructor(problemState) {
    const parserOptions = {
      ignoreAttributes: false,
      alwaysCreateTextNode: true,
    };
    const questionParserOptions = {
      ignoreAttributes: false,
      alwaysCreateTextNode: true,
      preserveOrder: true,
    };
    const questionBuilderOptions = {
      ignoreAttributes: false,
      attributeNamePrefix: '@_',
      suppressBooleanAttributes: false,
      format: true,
      preserveOrder: true,
    };
    const builderOptions = {
      ignoreAttributes: false,
      attributeNamePrefix: '@_',
      suppressBooleanAttributes: false,
      format: true,
    };
    this.questionParser = new XMLParser(questionParserOptions);
    this.parser = new XMLParser(parserOptions);
    this.builder = new XMLBuilder(builderOptions);
    this.questionBuilder = new XMLBuilder(questionBuilderOptions);
    this.editorObject = problemState.editorObject;
    this.problemState = problemState.problem;
  }

  addHints() {
    const hintsArray = [];
    const hints = _.get(this.problemState, 'settings.hints', []);
    hints.forEach(element => {
      hintsArray.push({
        '#text': element.value,
      });
    });
    const demandhint = {
      demandhint: {
        hint: hintsArray,
      },
    };
    return demandhint;
  }

  addSolution() {
    if (!_.has(this.problemState, 'settings.solutionExplanation')) { return {}; }
    const { solution } = this.editorObject;
    const parsedSolution = this.parser.parse(solution);
    const solutionObject = {
      solution: {
        ...parsedSolution,
      },
    };
    return solutionObject;
  }

  addMultiSelectAnswers(option) {
    const choice = [];
    let compoundhint = [];
    let widget = {};
    // eslint-disable-next-line prefer-const
    let { answers, generalFeedback } = this.problemState;
    const answerTitles = this.editorObject?.answers;
    // general feedback replaces selected feedback if all incorrect selected feedback is the same.
    if (generalFeedback !== ''
    && answers.every(
      answer => (
        answer.correct
          ? true
          : answer?.selectedFeedback === answers.find(a => a.correct === false).selectedFeedback
      ),
    )) {
      answers = answers.map(answer => (!answer?.correct
        ? { ...answer, selectedFeedback: generalFeedback }
        : answer));
    }
    answers.forEach((answer) => {
      const feedback = [];
      let singleAnswer = {};
      const title = answerTitles ? answerTitles[answer.id] : answer.title;
      if (this.hasAttributeWithValue(answer, 'title')) {
        if (this.hasAttributeWithValue(answer, 'selectedFeedback')) {
          feedback.push({
            '#text': _.get(answer, 'selectedFeedback'),
            '@_selected': true,
          });
        }
        if (this.hasAttributeWithValue(answer, 'unselectedFeedback')) {
          feedback.push({
            '#text': _.get(answer, 'unselectedFeedback'),
            '@_selected': false,
          });
        }
        if (this.hasAttributeWithValue(answer, 'feedback')) {
          feedback.push({
            '#text': _.get(answer, 'feedback'),
          });
        }
        if (feedback.length) {
          singleAnswer[`${option}hint`] = feedback;
        }
        singleAnswer = {
          '#text': title,
          '@_correct': answer.correct,
          ...singleAnswer,
        };
        choice.push(singleAnswer);
      }
    });
    widget = { [option]: choice };
    if (_.has(this.problemState, 'groupFeedbackList')) {
      compoundhint = this.addGroupFeedbackList();
      widget = {
        ...widget,
        compoundhint,
      };
    }
    return widget;
  }

  addGroupFeedbackList() {
    const compoundhint = [];
    const { groupFeedbackList } = this.problemState;
    groupFeedbackList.forEach((element) => {
      compoundhint.push({
        '#text': element.feedback,
        '@_value': element.answers.join(' '),
      });
    });
    return compoundhint;
  }

  addQuestion() {
    const question = this.editorObject?.question || this.problemState?.question;
    // const question = this.editorObject?.question || '';
    const questionObject = this.questionParser.parse(question);
    return questionObject;
  }

  buildMultiSelectProblem(problemType, widget, option) {
    const question = this.addQuestion();
    const widgetObject = this.addMultiSelectAnswers(option);
    const demandhint = this.addHints();
    const solution = this.addSolution();

    const problemObject = {
      problem: {
        [problemType]: {
          [widget]: widgetObject,
          ...solution,
        },
        ...demandhint,
      },
    };

    const problem = this.builder.build(problemObject);
    const questionString = this.questionBuilder.build(question);
    let problemTypeTag;
    switch (problemType) {
      case ProblemTypeKeys.MULTISELECT:
        [problemTypeTag] = problem.match(/<choiceresponse>|<choiceresponse.[^>]+>/);
        break;
      case ProblemTypeKeys.DROPDOWN:
        [problemTypeTag] = problem.match(/<optionresponse>|<optionresponse.[^>]+>/);
        break;
      case ProblemTypeKeys.SINGLESELECT:
        [problemTypeTag] = problem.match(/<multiplechoiceresponse>|<multiplechoiceresponse.[^>]+>/);
        break;
      default:
        break;
    }
    const updatedString = `${problemTypeTag}\n${questionString}`;
    const problemString = problem.replace(problemTypeTag, updatedString);

    return problemString;
  }

  buildTextInput() {
    const question = this.addQuestion();
    const demandhint = this.addHints();
    const answerObject = this.buildTextInputAnswersFeedback();
    const solution = this.addSolution();

    const problemObject = {
      problem: {
        [ProblemTypeKeys.TEXTINPUT]: {
          ...answerObject,
          ...solution,
        },
        ...demandhint,
      },
    };

    const problem = this.builder.build(problemObject);
    const questionString = this.questionBuilder.build(question);
    const [problemTypeTag] = problem.match(/<stringresponse>|<stringresponse.[^>]+>/);
    const updatedString = `${problemTypeTag}\n${questionString}`;
    const problemString = problem.replace(problemTypeTag, updatedString);

    return problemString;
  }

  buildTextInputAnswersFeedback() {
    const { answers } = this.problemState;
    const answerTitles = this.editorObject?.answers;
    let answerObject = {};
    const additionAnswers = [];
    const wrongAnswers = [];
    let firstCorrectAnswerParsed = false;
    answers.forEach((answer) => {
      const correcthint = this.getAnswerHints(answer);
      if (this.hasAttributeWithValue(answer, 'title')) {
        const title = answerTitles ? answerTitles[answer.id] : answer.title;
        if (answer.correct && firstCorrectAnswerParsed) {
          additionAnswers.push({
            '@_answer': title,
            ...correcthint,
          });
        } else if (answer.correct && !firstCorrectAnswerParsed) {
          firstCorrectAnswerParsed = true;
          answerObject = {
            '@_answer': title,
            ...correcthint,
          };
        } else if (!answer.correct) {
          wrongAnswers.push({
            '@_answer': title,
            '#text': answer.selectedFeedback,
          });
        }
      }
    });
    answerObject = {
      ...answerObject,
      additional_answer: additionAnswers,
      stringequalhint: wrongAnswers,
      '@_type': _.get(this.problemState, 'additionalAttributes.type', 'ci'),
      textline: {
        '@_size': _.get(this.problemState, 'additionalAttributes.textline.size', 20),
      },
    };
    return answerObject;
  }

  buildNumericInput() {
    const question = this.addQuestion();
    const demandhint = this.addHints();
    const answerObject = this.buildNumericalResponse();
    const solution = this.addSolution();

    const problemObject = {
      problem: {
        [ProblemTypeKeys.NUMERIC]: {
          ...answerObject,
          ...solution,
        },
        ...demandhint,
      },
    };

    const problem = this.builder.build(problemObject);
    const questionString = this.questionBuilder.build(question);
    const [problemTypeTag] = problem.match(/<numericalresponse>|<numericalresponse.[^>]+>/);
    const updatedString = `${questionString}\n${problemTypeTag}`;
    const problemString = problem.replace(problemTypeTag, updatedString);

    return problemString;
  }

  buildNumericalResponse() {
    const { answers } = this.problemState;
    const answerTitles = this.editorObject?.answers;
    let answerObject = {};
    const additionalAnswers = [];
    let firstCorrectAnswerParsed = false;
    answers.forEach((answer) => {
      const correcthint = this.getAnswerHints(answer);
      if (this.hasAttributeWithValue(answer, 'title')) {
        const title = answerTitles ? answerTitles[answer.id] : answer.title;
        if (answer.correct && !firstCorrectAnswerParsed) {
          firstCorrectAnswerParsed = true;
          let responseParam = {};
          if (_.has(answer, 'tolerance')) {
            responseParam = {
              responseparam: {
                '@_type': 'tolerance',
                '@_default': _.get(answer, 'tolerance', 0),
              },
            };
          }
          answerObject = {
            '@_answer': title,
            ...responseParam,
            ...correcthint,
          };
        } else if (answer.correct && firstCorrectAnswerParsed) {
          additionalAnswers.push({
            '@_answer': title,
            ...correcthint,
          });
        }
      }
    });
    answerObject = {
      ...answerObject,
      additional_answer: additionalAnswers,
      formulaequationinput: {
        '#text': '',
      },
    };
    return answerObject;
  }

  getAnswerHints(elementObject) {
    const feedback = elementObject?.selectedFeedback;
    let correcthint = {};
    if (feedback !== undefined && feedback !== '') {
      correcthint = {
        correcthint: {
          '#text': feedback,
        },
      };
    }
    return correcthint;
  }

  hasAttributeWithValue(obj, attr) {
    return _.has(obj, attr) && _.get(obj, attr, '').toString().trim() !== '';
  }

  buildOLX() {
    const { problemType } = this.problemState;
    let problemString = '';

    switch (problemType) {
      case ProblemTypeKeys.MULTISELECT:
        problemString = this.buildMultiSelectProblem(ProblemTypeKeys.MULTISELECT, 'checkboxgroup', 'choice');
        break;
      case ProblemTypeKeys.DROPDOWN:
        problemString = this.buildMultiSelectProblem(ProblemTypeKeys.DROPDOWN, 'optioninput', 'option');
        break;
      case ProblemTypeKeys.SINGLESELECT:
        problemString = this.buildMultiSelectProblem(ProblemTypeKeys.SINGLESELECT, 'choicegroup', 'choice');
        break;
      case ProblemTypeKeys.TEXTINPUT:
        problemString = this.buildTextInput();
        break;
      case ProblemTypeKeys.NUMERIC:
        problemString = this.buildNumericInput();
        break;
      default:
        break;
    }
    return problemString;
  }
}

export default ReactStateOLXParser;
