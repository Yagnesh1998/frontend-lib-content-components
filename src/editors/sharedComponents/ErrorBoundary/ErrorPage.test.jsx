import React from 'react';
import { shallow } from 'enzyme';
import { selectors } from '../../data/redux';
import { formatMessage } from '../../../testUtils';
import { ErrorPage, mapStateToProps } from './ErrorPage';

jest.mock('../../data/redux', () => ({
  selectors: {
    app: {
      unitUrl: jest.fn(state => ({ unitUrl: state })),
    },
  },
}));

describe('Editor Page', () => {
  const emptyProps = {
    courseId: null,
    studioEndpointUrl: null,
    intl: { formatMessage },
  };
  const passedProps = {
    courseId: 'course-v1:edX+DemoX+Demo_Course',
    studioEndpointUrl: 'fakeurl.com',
    intl: { formatMessage },
  };
  const unitData = {
    data: {
      ancestors: [{ id: 'SomeID' }],
    },
  };

  describe('rendered with empty props', () => {
    it('should only have one button (try again)', () => {
      const wrapper = shallow(<ErrorPage {...emptyProps} />);
      const buttonText = wrapper.find('Button').text();
      expect(wrapper).toMatchSnapshot();
      expect(buttonText).toEqual('Try again');
    });
  });

  describe('rendered with courseId and studioEndpointUrl defined', () => {
    it('should only have two buttons (return to studio and try again)', () => {
      const wrapper = shallow(<ErrorPage {...passedProps} />);
      const firstButtonText = wrapper.find('Button').at(0).text();
      const secondButtonText = wrapper.find('Button').at(1).text();
      expect(wrapper).toMatchSnapshot();
      expect(firstButtonText).toBeDefined();
      expect(secondButtonText).toEqual('Try again');
    });
  });
  test('rendered correctly with expected unitData defined', () => {
    expect(shallow(<ErrorPage {...passedProps} unitData={unitData} />)).toMatchSnapshot();
  });
  describe('mapStateToProps() function', () => {
    const testState = { A: 'pple', B: 'anana', C: 'ucumber' };
    test('unitData should equal unitUrl from app.unitUrl', () => {
      expect(
        mapStateToProps(testState).unitData,
      ).toEqual(selectors.app.unitUrl(testState));
    });
  });
});
