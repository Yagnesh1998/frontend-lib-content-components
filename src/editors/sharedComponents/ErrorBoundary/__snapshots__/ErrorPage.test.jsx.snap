// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor Page rendered correctly with expected unitData defined 1`] = `
<Container
  className="py-5 justify-content-center align-items-start text-center"
  fluid={true}
>
  <Row>
    <Col>
      <p
        className="text-muted"
      >
        An unexpected error occurred. Please click the button below to refresh the page.
      </p>
      <Row
        className="justify-content-center"
      >
        <Button
          className="mr-2"
          onClick={[Function]}
          variant="outline-primary"
        >
          {unitUrl, select,
      null {Return to course outline}
      other {Return to unit page}
    }
        </Button>
        <Button
          className="ml-2"
          onClick={[Function]}
        >
          Try again
        </Button>
      </Row>
    </Col>
  </Row>
</Container>
`;

exports[`Editor Page rendered with courseId and studioEndpointUrl defined should only have two buttons (return to studio and try again) 1`] = `
<Container
  className="py-5 justify-content-center align-items-start text-center"
  fluid={true}
>
  <Row>
    <Col>
      <p
        className="text-muted"
      >
        An unexpected error occurred. Please click the button below to refresh the page.
      </p>
      <Row
        className="justify-content-center"
      >
        <Button
          className="mr-2"
          onClick={[Function]}
          variant="outline-primary"
        >
          {unitUrl, select,
      null {Return to course outline}
      other {Return to unit page}
    }
        </Button>
        <Button
          className="ml-2"
          onClick={[Function]}
        >
          Try again
        </Button>
      </Row>
    </Col>
  </Row>
</Container>
`;

exports[`Editor Page rendered with empty props should only have one button (try again) 1`] = `
<Container
  className="py-5 justify-content-center align-items-start text-center"
  fluid={true}
>
  <Row>
    <Col>
      <p
        className="text-muted"
      >
        An unexpected error occurred. Please click the button below to refresh the page.
      </p>
      <Row
        className="justify-content-center"
      >
        <Button
          className="ml-2"
          onClick={[Function]}
        >
          Try again
        </Button>
      </Row>
    </Col>
  </Row>
</Container>
`;
